query-planner-js/src/buildPlan.ts

```diff
@@ -1326,183 +1210,583 @@ class FetchDependencyGraph {
+    for (const [g, edges] of r1) {
+      const newEdges = this.mergeRemaingsAndRemoveIfFound(g, edges, r2);
+      if (newEdges.length == 0) {
+        toHandle.push(g);
+      } else {
+        unhandled.push([g, newEdges])
+      }
+    }
+    unhandled.push(...r2);
+    return [toHandle, unhandled];
+  }
+
+  private mergeRemaingsAndRemoveIfFound(group: FetchGroup, inEdges: UnhandledParentRelations, otherGroups: UnhandledGroups): UnhandledParentRelations {
+    const idx = otherGroups.findIndex(g => g[0] === group);
+    if (idx < 0) {
+      return inEdges;
+    } else {
+      const otherEdges = otherGroups[idx][1];
+      otherGroups.splice(idx, 1);
+      // The uhandled are the one that are unhandled on both side.
+      return inEdges.filter(e => otherEdges.includes(e))
+    }
+  }
+
+  process<G, P>(processor: FetchGroupProcessor<G, P, any>): G[] {
+    this.reduceAndOptimize();
+
+    const rootNodes: G[] = this.rootGroups.values().map(rootGroup => {
+      const [node, remaining] = this.processGroup(processor, rootGroup, true);
+      assert(remaining.length == 0, () => `Root group ${rootGroup} should have no remaining groups unhandled, but got ${remaining}`);
+      return node;
+    });
+    return rootNodes;
+  }
+
+  dumpOnConsole(msg?: string) {
+    if (msg) {
+      console.log(msg);
+    }
+    console.log('Groups:');
+    for (const group of this.groups) {
+      console.log(`  ${group}`);
+    }
+    console.log('Children relationships:');
+    for (const group of this.groups) {
+      const children = group.children();
+      if (children.length === 1) {
+        console.log(`  [${group.index}] => [ ${children[0]} ]`);
+      } else if (children.length !== 0) {
+        console.log(`  [${group.index}] => [\n    ${children.join('\n    ')}\n  ]`);
+      }
+    }
+    console.log('Parent relationships:');
+    const printParentRelation = (rel: ParentRelation) => (
+      rel.path ? `${rel.group} (path: [${rel.path.join(', ')}])` : rel.group.toString()
+    );
+    for (const group of this.groups) {
+      const parents = group.parents();
+      if (parents.length === 1) {
+        console.log(`  [${group.index}] => [ ${printParentRelation(parents[0])} ]`);
+      } else if (parents.length !== 0) {
+        console.log(`  [${group.index}] => [\n    ${parents.map(printParentRelation).join('\n    ')}\n  ]`);
+      }
+    }
+    console.log('--------');
+  }
+
+  toString() : string {
+    return this.rootGroups.values().map(g => this.toStringInternal(g, "")).join('\n');
+  }
+
+  toStringInternal(group: FetchGroup, indent: string): string {
+    const children = group.children();
+    return [indent + group.subgraphName + ' <- ' + children.map((child) => child.subgraphName).join(', ')]
+      .concat(children
+        .flatMap(g => g.children().length == 0
+          ? []
+          : this.toStringInternal(g, indent + "  ")))
+      .join('\n');
+  }
+}
+
+interface FetchGroupProcessor<G, P, F> {
```

Comment:
Can you name the templatized types to convey more meaning?

Reply:
I did, though had a hard to come up with very short names, so hopefully it's an improvement.